File: lib/java/opentoken/src/main/java/com/truveta/opentoken/tokens/TokenGenerator.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.tokens;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import lombok.Getter;
import lombok.Setter;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.truveta.opentoken.attributes.Attribute;
import com.truveta.opentoken.attributes.AttributeExpression;
import com.truveta.opentoken.attributes.AttributeLoader;
import com.truveta.opentoken.attributes.person.BirthDateAttribute;
import com.truveta.opentoken.attributes.person.FirstNameAttribute;
import com.truveta.opentoken.attributes.person.LastNameAttribute;
import com.truveta.opentoken.attributes.person.PostalCodeAttribute;
import com.truveta.opentoken.attributes.person.SexAttribute;
import com.truveta.opentoken.tokens.tokenizer.SHA256Tokenizer;
import com.truveta.opentoken.tokens.tokenizer.Tokenizer;
import com.truveta.opentoken.tokens.tokenizer.PassthroughTokenizer;
import com.truveta.opentoken.tokentransformer.TokenTransformer;

/**
 * Generates both the token signature and the token itself.
 */
@Getter
@Setter
public class TokenGenerator implements Serializable {
    private static final long serialVersionUID = 1L;
    private static final transient Logger logger = LoggerFactory.getLogger(TokenGenerator.class);
    private static final String T6_RULE_ID = "T6";

    private Tokenizer tokenizer;
    private BaseTokenDefinition tokenDefinition;

    private Map<Class<? extends Attribute>, Attribute> attributeInstanceMap;

    /**
     * Initializes the token generator.
     * 
     * @param tokenDefinition      the token definition.
     * @param tokenTransformerList a list of token transformers.
     * @deprecated Use {@link #TokenGenerator(BaseTokenDefinition, Tokenizer)} instead.
     *             This constructor will be removed in a future release.
     */
    @Deprecated(since = "1.12.0", forRemoval = true)
    public TokenGenerator(BaseTokenDefinition tokenDefinition, List<TokenTransformer> tokenTransformerList) {
        this(tokenDefinition, new SHA256Tokenizer(tokenTransformerList));
    }

    /**
     * Initializes the token generator with an explicit tokenizer.
     *
     * @param tokenDefinition      the token definition.
     * @param tokenizer            optional tokenizer implementation. Use
     *                             {@link PassthroughTokenizer} for plain mode.
     */
    public TokenGenerator(BaseTokenDefinition tokenDefinition, Tokenizer tokenizer) {
        this.tokenDefinition = tokenDefinition;
        this.attributeInstanceMap = new HashMap<>();
        AttributeLoader.load().forEach(attribute -> attributeInstanceMap.put(attribute.getClass(), attribute));
        this.tokenizer = tokenizer;
    }

    /*
     * Get the token signature for a given token identifier. Populates the
     * invalidAttributes list in the result object with the attributes that are
     * invalid.
     *
     * @param tokenId the token identifier.
     * 
     * @param personAttributes The person attributes. It is a map of the person
     * attributes.
     * 
     * @param result the token generator result.
     * 
     * @return the token signature using the token definition for the given token
     * identifier.
     */
    protected String getTokenSignature(String tokenId, Map<Class<? extends Attribute>, String> personAttributes,
            TokenGeneratorResult result) {
        if (T6_RULE_ID.equals(tokenId)) {
            return getT6Signature(personAttributes, result);
        }

        var definition = tokenDefinition.getTokenDefinition(tokenId);
        if (personAttributes == null) {
            throw new IllegalArgumentException("Person attributes cannot be null.");
        }
        if (definition == null) {
            return null;
        }

        var values = new ArrayList<String>(definition.size());

        for (AttributeExpression attributeExpression : definition) {
            if (!personAttributes.containsKey(attributeExpression.getAttributeClass())) {
                return null;
            }

            var attribute = attributeInstanceMap.get(attributeExpression.getAttributeClass());
            String attributeValue = personAttributes.get(attributeExpression.getAttributeClass());
            if (!attribute.validate(attributeValue)) {
                result.getInvalidAttributes().add(attribute.getName());
                return null;
            }

            attributeValue = attribute.normalize(attributeValue);

            try {
                attributeValue = attributeExpression.getEffectiveValue(attributeValue);
                values.add(attributeValue);
            } catch (IllegalArgumentException e) {
                logger.error(e.getMessage());
                return null;
            }

        }

        return Stream.of(values.toArray(new String[0])).filter(s -> null != s && !s.isBlank())
                .collect(Collectors.joining("|"));
    }

    private String getT6Signature(Map<Class<? extends Attribute>, String> personAttributes,
            TokenGeneratorResult result) {
        if (!T6InferenceConfig.isEnabled()) {
            return null;
        }

        String payloadJson = buildT6Payload(personAttributes, result);
        if (payloadJson == null) {
            return null;
        }

        try {
            return T6OnnxSignatureGenerator.generateSignature(payloadJson);
        } catch (Exception e) {
            logger.error("Error generating token signature for token id: {}", T6_RULE_ID, e);
            return null;
        }
    }

    public String buildT6Payload(Map<Class<? extends Attribute>, String> personAttributes,
            TokenGeneratorResult result) {
        if (personAttributes == null) {
            return null;
        }

        Map<String, String> payload = new LinkedHashMap<>();
        if (!addT6Field(PostalCodeAttribute.class, "PostalCode", personAttributes, result, payload)
                || !addT6Field(BirthDateAttribute.class, "Birthdate", personAttributes, result, payload)
                || !addT6Field(FirstNameAttribute.class, "GivenName", personAttributes, result, payload)
                || !addT6Field(LastNameAttribute.class, "Surname", personAttributes, result, payload)
                || !addT6Field(SexAttribute.class, "Gender", personAttributes, result, payload)) {
            return null;
        }

        return asJson(payload);
    }

    public TokenGeneratorResult getAllTokensExcludingT6(Map<Class<? extends Attribute>, String> personAttributes) {
        TokenGeneratorResult result = new TokenGeneratorResult();

        for (String tokenId : tokenDefinition.getTokenIdentifiers()) {
            if (T6_RULE_ID.equals(tokenId)) {
                continue;
            }
            try {
                var token = getToken(tokenId, personAttributes, result);
                if (token != null) {
                    result.getTokens().put(tokenId, token);
                }
            } catch (Exception e) {
                logger.error("Error generating token for token id: " + tokenId, e);
            }
        }

        return result;
    }

    public void applyT6SignatureToken(TokenGeneratorResult result, String signature) {
        try {
            String token = tokenizer.tokenize(signature);
            result.getTokens().put(T6_RULE_ID, token);
            if (Token.BLANK.equals(token)) {
                result.getBlankTokensByRule().add(T6_RULE_ID);
            }
        } catch (Exception e) {
            logger.error("Error generating token for token id: " + T6_RULE_ID, e);
            result.getTokens().put(T6_RULE_ID, Token.BLANK);
            result.getBlankTokensByRule().add(T6_RULE_ID);
        }
    }

    private boolean addT6Field(Class<? extends Attribute> attributeClass, String fieldName,
            Map<Class<? extends Attribute>, String> personAttributes, TokenGeneratorResult result,
            Map<String, String> payload) {
        if (!personAttributes.containsKey(attributeClass)) {
            return false;
        }

        Attribute attribute = attributeInstanceMap.get(attributeClass);
        if (attribute == null) {
            return false;
        }

        String value = personAttributes.get(attributeClass);
        if (!attribute.validate(value)) {
            result.getInvalidAttributes().add(attribute.getName());
            return false;
        }

        String normalized = attribute.normalize(value);
        if (normalized == null || normalized.isBlank()) {
            result.getInvalidAttributes().add(attribute.getName());
            return false;
        }

        payload.put(fieldName, normalized);
        return true;
    }

    private String asJson(Map<String, String> values) {
        StringBuilder builder = new StringBuilder("{");
        boolean first = true;
        for (Map.Entry<String, String> entry : values.entrySet()) {
            if (!first) {
                builder.append(", ");
            }
            builder.append("\"").append(escapeJson(entry.getKey())).append("\": ");
            builder.append("\"").append(escapeJson(entry.getValue())).append("\"");
            first = false;
        }
        builder.append("}");
        return builder.toString();
    }

    private String escapeJson(String value) {
        return value.replace("\\", "\\\\").replace("\"", "\\\"");
    }

    /**
     * Get the token signatures for all token/rule identifiers. This is mostly a
     * debug/logging/test method.
     * 
     * @param personAttributes the person attributes map.
     * 
     * @return A map of token/rule identifier to the token signature.
     */
    public Map<String, String> getAllTokenSignatures(Map<Class<? extends Attribute>, String> personAttributes) {
        var signatures = new HashMap<String, String>();
        for (String tokenId : tokenDefinition.getTokenIdentifiers()) {
            try {
                var signature = getTokenSignature(tokenId, personAttributes, new TokenGeneratorResult());
                if (signature != null) {
                    signatures.put(tokenId, signature);
                }
            } catch (Exception e) {
                logger.error("Error generating token signature for token id: " + tokenId, e);
            }
        }
        return signatures;
    }

    /*
     * Get token for a given token identifier.
     *
     * @param tokenId the token identifier.
     * 
     * @param personAttributes the person attributes map.
     * 
     * @param result the token generator result.
     * 
     * @return the token using the token definition for the given token identifier.
     * 
     * @throws TokenGenerationException in case of failure to generate the token.
     */
    protected String getToken(String tokenId, Map<Class<? extends Attribute>, String> personAttributes,
            TokenGeneratorResult result)
            throws TokenGenerationException {
        var signature = getTokenSignature(tokenId, personAttributes, result);
        logger.debug("Token signature for token id {}: {}", tokenId, signature);
        try {
            String token = tokenizer.tokenize(signature);
            // Track blank tokens by rule
            if (Token.BLANK.equals(token)) {
                result.getBlankTokensByRule().add(tokenId);
            }
            return token;
        } catch (Exception e) {
            logger.error("Error generating token for token id: " + tokenId, e);
            throw new TokenGenerationException("Error generating token", e);
        }
    }

    /**
     * Get the tokens for all token/rule identifiers.
     * 
     * @param personAttributes the person attributes map.
     * 
     * @return A {@link TokenGeneratorResult} object containing the tokens and
     *         invalid attributes.
     */
    public TokenGeneratorResult getAllTokens(Map<Class<? extends Attribute>, String> personAttributes) {
        TokenGeneratorResult result = new TokenGeneratorResult();

        for (String tokenId : tokenDefinition.getTokenIdentifiers()) {
            try {
                var token = getToken(tokenId, personAttributes, result);
                if (token != null) {
                    result.getTokens().put(tokenId, token);
                }
            } catch (Exception e) {
                logger.error("Error generating token for token id: " + tokenId, e);
            }
        }

        return result;
    }

    /**
     * Get invalid person attribute names.
     * 
     * @param personAttributes the person attributes map.
     * 
     * @return A set of invalid person attribute names.
     */
    public Set<String> getInvalidPersonAttributes(Map<Class<? extends Attribute>, String> personAttributes) {
        var response = new HashSet<String>();

        for (Map.Entry<Class<? extends Attribute>, String> entry : personAttributes.entrySet()) {
            if (!attributeInstanceMap.get(entry.getKey()).validate(entry.getValue())) {
                response.add(attributeInstanceMap.get(entry.getKey()).getName());
            }
        }

        return response;
    }
}

File: lib/python/opentoken/src/main/opentoken/tokens/token_generator.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import logging
from importlib.metadata import entry_points
from typing import Dict, List, Optional, Set, Type

from opentoken.attributes.attribute import Attribute
from opentoken.attributes.attribute_loader import AttributeLoader
from opentoken.tokens.base_token_definition import BaseTokenDefinition
from opentoken.tokens.inference_signature_provider import InferenceBatchResult, InferenceSignatureProvider  # noqa: F401
from opentoken.tokens.token import Token
from opentoken.tokens.token_generation_exception import TokenGenerationException
from opentoken.tokens.token_generator_result import TokenGeneratorResult
from opentoken.tokens.tokenizer.sha256_tokenizer import SHA256Tokenizer
from opentoken.tokens.tokenizer.tokenizer import Tokenizer
from opentoken.tokentransformer.token_transformer import TokenTransformer

logger = logging.getLogger(__name__)

_inference_provider: Optional[InferenceSignatureProvider] = None
_provider_discovered = False


def _get_inference_provider() -> Optional[InferenceSignatureProvider]:
    """Lazily discover and cache the first registered InferenceSignatureProvider."""
    global _inference_provider, _provider_discovered
    if not _provider_discovered:
        _provider_discovered = True
        eps = entry_points(group="opentoken.inference_providers")
        for ep in eps:
            try:
                provider_cls = ep.load()
                _inference_provider = provider_cls()
                break
            except Exception:
                pass
    return _inference_provider


class TokenGenerator:
    """Generates both the token signature and the token itself."""

    @classmethod
    def from_transformers(
        cls, token_definition: BaseTokenDefinition, token_transformer_list: List[TokenTransformer]
    ) -> "TokenGenerator":
        """
        Convenience constructor that creates a TokenGenerator with SHA256Tokenizer.

        Args:
            token_definition: The token definition.
            token_transformer_list: A list of token transformers.

        Returns:
            A TokenGenerator instance with SHA256Tokenizer.
        """
        return cls(token_definition, SHA256Tokenizer(token_transformer_list))

    def __init__(self, token_definition: BaseTokenDefinition, tokenizer: Tokenizer):
        """
        Initialize the token generator with an explicit tokenizer.

        Args:
            token_definition: The token definition.
            tokenizer: Tokenizer implementation. Use PassthroughTokenizer for plain mode.
        """
        self.token_definition = token_definition
        self.attribute_instance_map: Dict[Type[Attribute], Attribute] = {}

        # Load attributes
        for attribute in AttributeLoader.load():
            self.attribute_instance_map[type(attribute)] = attribute

        self.tokenizer = tokenizer

    def _get_token_signature(
        self, token_id: str, person_attributes: Dict[Type[Attribute], str], result: TokenGeneratorResult
    ) -> Optional[str]:
        """
        Get the token signature for a given token identifier.

        Populates the invalid_attributes list in the result object with the attributes
        that are invalid.

        Args:
            token_id: The token identifier.
            person_attributes: The person attributes map.
            result: The token generator result.

        Returns:
            The token signature using the token definition for the given token identifier.
        """
        provider = _get_inference_provider()
        if provider is not None and provider.get_token_id() == token_id and provider.is_enabled():
            try:
                return provider.generate_signature(person_attributes)
            except Exception as error:
                logger.error("Error generating token signature for token id: %s", token_id, exc_info=error)
                return None

        definition = self.token_definition.get_token_definition(token_id)

        if person_attributes is None:
            raise ValueError("Person attributes cannot be null.")

        if definition is None:
            return None

        values = []

        for attribute_expression in definition:
            attribute_class = attribute_expression.attribute_class

            if attribute_class not in person_attributes:
                return None

            attribute = self.attribute_instance_map.get(attribute_class)
            if attribute is None:
                return None

            attribute_value = person_attributes[attribute_class]

            if not attribute.validate(attribute_value):
                result.invalid_attributes.add(attribute.get_name())
                return None

            attribute_value = attribute.normalize(attribute_value)

            try:
                attribute_value = attribute_expression.get_effective_value(attribute_value)
                values.append(attribute_value)
            except ValueError as e:
                logger.error(str(e))
                return None

        # Filter out None and blank values, then join with '|'
        filtered_values = [v for v in values if v is not None and v.strip() != ""]
        return "|".join(filtered_values)

    def get_all_token_signatures(self, person_attributes: Dict[Type[Attribute], str]) -> Dict[str, str]:
        """
        Get the token signatures for all token/rule identifiers.

        This is mostly a debug/logging/test method.

        Args:
            person_attributes: The person attributes map.

        Returns:
            A map of token/rule identifier to the token signature.
        """
        signatures = {}

        for token_id in self.token_definition.get_token_identifiers():
            try:
                signature = self._get_token_signature(token_id, person_attributes, TokenGeneratorResult())
                if signature is not None:
                    signatures[token_id] = signature
            except Exception as e:
                logger.error(f"Error generating token signature for token id: {token_id}", exc_info=e)

        return signatures

    def _get_token(
        self, token_id: str, person_attributes: Dict[Type[Attribute], str], result: TokenGeneratorResult
    ) -> Optional[str]:
        """
        Get token for a given token identifier.

        Args:
            token_id: The token identifier.
            person_attributes: The person attributes map.
            result: The token generator result.

        Returns:
            The token using the token definition for the given token identifier.

        Raises:
            TokenGenerationException: In case of failure to generate the token.
        """
        signature = self._get_token_signature(token_id, person_attributes, result)
        logger.debug(f"Token signature for token id {token_id}: {signature}")

        try:
            token = self.tokenizer.tokenize(signature)
            # Track blank tokens by rule
            if Token.BLANK == token:
                result.blank_tokens_by_rule.add(token_id)
            return token
        except Exception as e:
            logger.error(f"Error generating token for token id: {token_id}", exc_info=e)
            raise TokenGenerationException("Error generating token", e)

    def get_all_tokens(self, person_attributes: Dict[Type[Attribute], str]) -> TokenGeneratorResult:
        """
        Get the tokens for all token/rule identifiers.

        Args:
            person_attributes: The person attributes map.

        Returns:
            A TokenGeneratorResult object containing the tokens and invalid attributes.
        """
        result = TokenGeneratorResult()

        for token_id in self.token_definition.get_token_identifiers():
            try:
                token = self._get_token(token_id, person_attributes, result)
                if token is not None:
                    result.tokens[token_id] = token
            except Exception as e:
                logger.error(f"Error generating token for token id: {token_id}", exc_info=e)

        return result

    def generate_tokens_excluding(
        self,
        person_attributes: Dict[Type[Attribute], str],
        excluded_token_ids: Set[str],
    ) -> TokenGeneratorResult:
        """Get tokens for all rules except those in *excluded_token_ids*.

        Args:
            person_attributes: The person attributes map.
            excluded_token_ids: Set of token identifiers to skip.

        Returns:
            A TokenGeneratorResult object containing the tokens and invalid attributes.
        """
        result = TokenGeneratorResult()

        for token_id in self.token_definition.get_token_identifiers():
            if token_id in excluded_token_ids:
                continue
            try:
                token = self._get_token(token_id, person_attributes, result)
                if token is not None:
                    result.tokens[token_id] = token
            except Exception as error:
                logger.error(f"Error generating token for token id: {token_id}", exc_info=error)

        return result

    def apply_precomputed_signature(
        self, result: TokenGeneratorResult, token_id: str, signature: Optional[str]
    ) -> None:
        """Tokenize and store a token from a precomputed signature.

        Args:
            result: The token generator result to update.
            token_id: The token identifier key to store the result under.
            signature: The precomputed signature string, or ``None`` for a blank token.
        """
        try:
            token = self.tokenizer.tokenize(signature)
            result.tokens[token_id] = token
            if Token.BLANK == token:
                result.blank_tokens_by_rule.add(token_id)
        except Exception as error:
            logger.error("Error generating token for token id: %s", token_id, exc_info=error)
            result.tokens[token_id] = Token.BLANK
            result.blank_tokens_by_rule.add(token_id)

    def store_raw_token(self, result: TokenGeneratorResult, token_id: str, token_value: Optional[str]) -> None:
        """Store a pre-computed token value directly, bypassing the tokenizer and all transformers.

        Use for tokens whose value is already in its final form (e.g., internally
        pre-hashed signatures such as T6 rotation values).

        Args:
            result: The token generator result to update.
            token_id: The token identifier key to store the result under.
            token_value: The final token value, or ``None`` / blank to record a blank token.
        """
        if token_value and token_value != Token.BLANK:
            result.tokens[token_id] = token_value
        else:
            result.tokens[token_id] = Token.BLANK
            result.blank_tokens_by_rule.add(token_id)

    def apply_embedding_derived_tokens(
        self,
        result: TokenGeneratorResult,
        token_id_prefix: str,
        token_strings: List[str],
    ) -> None:
        """Apply pre-computed embedding-derived tokens to the result.

        Args:
            result: The token generator result to update.
            token_id_prefix: Prefix for derived token keys (e.g. ``"T6-R"``).
            token_strings: Pre-computed token strings from the embedding transformer.
        """
        for i, token_string in enumerate(token_strings):
            result.tokens[f"{token_id_prefix}{i}"] = token_string

    @staticmethod
    def get_inference_provider() -> Optional[InferenceSignatureProvider]:
        """Return the discovered InferenceSignatureProvider, or None if none is installed."""
        return _get_inference_provider()

    def get_invalid_person_attributes(self, person_attributes: Dict[Type[Attribute], str]) -> Set[str]:
        """
        Get invalid person attribute names.

        Args:
            person_attributes: The person attributes map.

        Returns:
            A set of invalid person attribute names.
        """
        response = set()

        for attribute_class, value in person_attributes.items():
            attribute = self.attribute_instance_map.get(attribute_class)
            if attribute and not attribute.validate(value):
                response.add(attribute.get_name())

        return response
